RecoLocalMuon/GEMRecHit/src/ME0SegAlgoMM.cc

```diff
@@ -0,0 +1,556 @@
+/**
+ * \file ME0SegAlgoMM.cc
+ *
+ *  \authors: Marcello Maggi
+ */
+ 
+#include "ME0SegAlgoMM.h"
+#include "DataFormats/GeometryVector/interface/GlobalPoint.h"
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+
+#include "CommonTools/Statistics/interface/ChiSquaredProbability.h"
+
+#include <algorithm>
+#include <cmath>
+#include <iostream>
+#include <string>
+
+/* Constructor
+ *
+ */
+ME0SegAlgoMM::ME0SegAlgoMM(const edm::ParameterSet& ps) : ME0SegmentAlgorithm(ps), myName("ME0SegAlgoMM") {
+	 
+  debug                     = ps.getUntrackedParameter<bool>("ME0Debug");
+  minHitsPerSegment         = ps.getParameter<unsigned int>("minHitsPerSegment");
+  preClustering             = ps.getParameter<bool>("preClustering");
+  dXclusBoxMax              = ps.getParameter<double>("dXclusBoxMax");
+  dYclusBoxMax              = ps.getParameter<double>("dYclusBoxMax");
+  preClustering_useChaining = ps.getParameter<bool>("preClusteringUseChaining");
+  dPhiChainBoxMax           = ps.getParameter<double>("dPhiChainBoxMax");
+  dEtaChainBoxMax           = ps.getParameter<double>("dEtaChainBoxMax");
+  maxRecHitsInCluster       = ps.getParameter<int>("maxRecHitsInCluster");
+}
+
+/* Destructor
+ *
+ */
+ME0SegAlgoMM::~ME0SegAlgoMM() {
+}
+
+
+std::vector<ME0Segment> ME0SegAlgoMM::run(ME0Ensamble ensamble, const EnsambleHitContainer& rechits) {
+
+  theEnsamble = ensamble;
```

Comment:
Please check that ensamble is valid before assigning it to theEnsamble -- in particular, the "first" element must not be null. We need to protect against a bad value getting into theEnsamble because it is used extensively in other methods.